experiment/cherrypicker/server.go

```diff
@@ -0,0 +1,196 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"regexp"
+	"time"
+
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/hook"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "cherrypick"
+
+var cherryPickRe = regexp.MustCompile(`(?m)^/cherrypick\s+(.+)$`)
+
+// Server implements http.Handler. It validates incoming GitHub webhooks and
+// then dispatches them to the appropriate plugins.
+type Server struct {
+	BotCredentials string
+	BotName        string
+	GitClient      *git.Client
+	GithubClient   *github.Client
+	HMACSecret     []byte
+	Logger         *logrus.Entry
+}
+
+// ServeHTTP validates an incoming webhook and puts it into the event channel.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	eventType, _, payload, ok := hook.ValidateWebhook(w, r, s.HMACSecret)
+	if !ok {
+		return
+	}
+	fmt.Fprint(w, "Event received. Have a nice day.")
+
+	// Handle only issue comments
+	if eventType != "issue_comment" {
+		return
+	}
+
+	var ic github.IssueCommentEvent
+	if err := json.Unmarshal(payload, &ic); err != nil {
+		s.Logger.WithError(err).Info("Unmarshaling of Github payload failed.")
+		return
+	}
+
+	go func() {
+		if err := s.handle(ic); err != nil {
+			s.Logger.WithError(err).Info("Cherry-pick failed.")
+		}
+	}()
+}
+
+func (s *Server) handle(ic github.IssueCommentEvent) error {
+	// Only consider new comments in closed PRs.
+	if !ic.Issue.IsPullRequest() || ic.Issue.State != "closed" || ic.Action != github.IssueCommentActionCreated {
+		return nil
+	}
+
+	cherryPickMatches := cherryPickRe.FindAllStringSubmatch(ic.Comment.Body, -1)
+	if len(cherryPickMatches) == 0 || len(cherryPickMatches[0]) != 2 {
+		return nil
+	}
+	targetBranch := cherryPickMatches[0][1]
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	commentAuthor := ic.Comment.User.Login
+
+	// TODO: Filter out notifications from repos we don't want to do cherry-picks.
+
+	pr, err := s.GithubClient.GetPullRequest(org, repo, num)
+	if err != nil {
+		return err
+	}
+	baseBranch := pr.Base.Ref
+
+	// TODO: Use a whitelist for allowed base and target branches.
+	if baseBranch != "master" || baseBranch == targetBranch {
+		resp := fmt.Sprintf("Base branch (%s) needs to be 'master' and differ from target branch (%s)", baseBranch, targetBranch)
+		s.Logger.Info(resp)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Only members should be able to do cherry-picks.
+	ok, err := s.GithubClient.IsMember(org, commentAuthor)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		resp := fmt.Sprintf("only %s org members may request cherry picks", org)
```

Comment:
I don't want to link any document here. I guess I can extend the message to hint that doing this manually is still acceptable but this command is meant to be used by authorized members only (at least initially).

Reply:
yeah non-members may try this way to cherry-pick.

Reply:
The issue is that we are going to use this downstream and I am not sure it will ever be used in kubernetes/kubernetes. That's why I want to avoid linking to any k8s-related doc. Extending the comment to inform them that they may still do it manually is doable.

Reply:
For myself, I want to see it in k/k :)